[functorch] Add some C++ documentation

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84648
* #84642
* #84639
* #84638
* __->__ #84637

Clarify the purpose of many files.

Differential Revision: [D39315053](https://our.internmc.facebook.com/intern/diff/D39315053)

zou3519 committed:
[functorch] Add some C++ documentation

Clarify the purpose of many files.

[ghstack-poisoned]

zou3519 committed:
Update on "[functorch] Add some C++ documentation"


Clarify the purpose of many files.

[ghstack-poisoned]

zou3519 committed:
Update on "[functorch] Add some C++ documentation"


Clarify the purpose of many files.

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39315053).

samdow commented on functorch/functorch/csrc/DynamicLayer.h:
```diff
@@ -18,12 +18,28 @@
 namespace functorch {
 
+// This file contains the implementation of functorch's interpreter stack.
+// See NOTE: [functorch interpreter stack] first before reading on.
```
Is this note already in the code? I'm just seeing it referenced a lot of places without seeing it, so wanted to make sure it wasn't forgotten about

zou3519 replied:
Yes: https://github.com/pytorch/pytorch/blob/31ef8ddb8c4467f5b8698ef1eb9bb8bab7056855/functorch/functorch/csrc/Interpreter.h#L12

zou3519 committed:
Update on "[functorch] Add some C++ documentation"


Clarify the purpose of many files.

Differential Revision: [D39315053](https://our.internmc.facebook.com/intern/diff/D39315053)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39315053).

zou3519 committed:
Update on "[functorch] Add some C++ documentation"


Clarify the purpose of many files.

Differential Revision: [D39315053](https://our.internmc.facebook.com/intern/diff/D39315053)

[ghstack-poisoned]

zou3519 committed:
Update on "[functorch] Add some C++ documentation"


Clarify the purpose of many files.

Differential Revision: [D39315053](https://our.internmc.facebook.com/intern/diff/D39315053)

[ghstack-poisoned]
